Add support for appending data to external tables - CSV

mustafasrepo opened the pull request:
# Which issue does this PR close?

Closes #

# Rationale for this change

This PR adds the support for the following SQL queries:

```sql
CREATE EXTERNAL TABLE source_table (
    a1  VARCHAR NOT NULL,
    a2  INT NOT NULL
)
STORED AS CSV
WITH HEADER ROW
OPTIONS ('UNBOUNDED' 'TRUE')
LOCATION '{source}';

CREATE EXTERNAL TABLE sink_table (
    a1  VARCHAR NOT NULL,
    a2  INT NOT NULL
)
STORED AS CSV
WITH HEADER ROW
OPTIONS ('UNBOUNDED' 'TRUE')
LOCATION '{sink}';

INSERT INTO sink_table
SELECT a1, a2 FROM source_table;
```

This PR adds support for appending data to external tables, which previously only supported memory tables. It introduces new structs and modifications to existing structs, enabling users to efficiently work with file-based storage systems when appending data.

# What changes are included in this PR?

- Added `FileSinkConfig` struct for base configurations when creating a physical plan for any given file format.
- Added `FileWriterExt` and to handle writing record batches to a file-like output.
- Added `CsvSink` struct for which implements `DataSink`  to write results to CSV file.

# Are these changes tested?

Yes

# Are there any user-facing changes?

This change allows users to append data to external tables, which was not possible before. Users can now work with file-based storage systems more efficiently, especially when appending data.

ozankabak commented:
I have been looking forward to this for a while. @metesynnada did a great job starting this work and @mustafasrepo took it to the finish line!

alamb commented on datafusion/core/src/physical_plan/file_format/csv.rs:
```diff
@@ -870,4 +948,66 @@ mod tests {
+    }
+
+    #[tokio::test]
+    async fn test_listing_table_insert_into() -> Result<()> {
```
I recommend this be put into an `sqllogictest` rather than here (as it appears to be a simple explain plan test)

alamb commented on datafusion/core/src/datasource/file_format/csv.rs:
```diff
@@ -324,6 +349,188 @@ fn build_schema_helper(names: Vec<String>, types: &[HashSet<DataType>]) -> Schem
+            Err(e) => {
+                // Abort all writers before returning the error:
+                for writer in $writers {
+                    let abort_future = writer.abort_writer()?;
```
I think this could be a function (which would make it easier to work with). Maybe like

```rust
fn check_for_errors(result: Result<RecordBatch>) -> Result<RecordBatch> {
...
}
```

Also, reading it I am not sure about the error handling. Specifically, 
1.  if the abort fails, the abort error may be reported rather then the initial write error.
2. If canceling one writer fails, the others may not be aborted (as `?` returns early) 

I have some ideas on how to refactor this (into a struct) and to make it more testable,  which I will try and put up as a draft PR shortly

mustafasrepo replied:
I converted macro, to function. Also during abortion we receive an error, that case doesn't return early. We try to abort all writers

alamb commented on datafusion/core/src/datasource/file_format/mod.rs:
```diff
@@ -87,6 +98,277 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
+/// `AsyncPutWriter` is an object that facilitates asynchronous writing to object stores.
+/// It is specifically designed for the `object_store` crate's `put` method and sends
+/// whole bytes at once when the buffer is flushed.
+pub struct AsyncPutWriter {
```
@tustvold  and/or @crepererum  (my go to people for rust async / stream expertise) I wonder if you have some time to review this code that adapts DataFusion to use the object_store put multi part features to do streaming writes

Pretty exciting!

tustvold replied:
I'm confused to see this here, we already provide AsynWrite from ObjectStore, so I'm not sure why we are re-implementing the buffering here? Am i missing something?

metesynnada replied:
We hypothesize that the consistent use of `put_multipart` for every put operation might adversely impact the cloud side, as it anticipates files exceeding a specific size (for example, 5MB for AWS). To mitigate this, we've developed a wrapper for the `put` operation that standardizes the write operation on `AsyncWrite`.

tustvold replied:
Oh yes 100%, my suggestion was be to type-erase at the level of the BatchSerializer and then have different impls for the different write modes. The async Read + abort interface feels a tad over complicated, at least imo.

alamb reviewed (approved):
Thank you @mustafasrepo  and @metesynnada and @ozankabak  -- this is really nice

I went though it carefully and while I had some small suggestions I also think this could be merged as is and iterate on it from there. 

Prior to review, I did not appreciate this PR lays the foundation for streaming (multi-part) writes to object store ❤️ 

I think the `FileSInkConfig` is an excellent idea and  follows the existing `FileScanConfig` pattern 👍  It also seems able to extend nicely to writing multiple file partitions (eventually) which is excellent

Here is a list of potential follow on work (some/all of which I would like to help with) -- If you agree, I can file ticket and help with this too as we would love to have streaming write support in IOx as well.  
- [ ] End to end sql level tests in sqllogictests (I can do this)
- [ ] Tests for streaming writes
- [ ] Tests for abort behavior (making sure all writers are canceled and the correct error is returned)
- [ ] Implement something similar for JSON and parquet. 
- [ ] Documentation / note that we support streaming multi-part writes

ozankabak commented:
> Here is a list of potential follow on work (some/all of which I would like to help with) -- If you agree, I can file ticket and help with this too as we would love to have streaming write support in IOx as well.

This sounds great! I checked in the comment improvements per your suggestions, @metesynnada and/or @mustafasrepo will shortly go through your other points. Thanks for the review.

ozankabak commented on datafusion/core/src/datasource/listing/table.rs:
```diff
@@ -759,6 +759,62 @@ impl TableProvider for ListingTable {
+
+        if self.table_paths().len() > 1 {
+            return Err(DataFusionError::Plan(
+                "Datafusion currently supports tables from single file.".to_owned(),
```
Done 👍

ozankabak commented on datafusion/core/src/datasource/listing/table.rs:
```diff
@@ -759,6 +759,62 @@ impl TableProvider for ListingTable {
+        // Get the object store for the table path.
+        let store = state.runtime_env().object_store(table_path)?;
+
+        let file_list_future = pruned_partition_list(
```
Agreed, done

ozankabak commented on datafusion/core/src/datasource/file_format/mod.rs:
```diff
@@ -87,6 +98,277 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
+    }
+}
+
+/// A simple wrapper around an `AsyncWrite` type.
```
Done, 👍

ozankabak commented on datafusion/core/src/physical_plan/file_format/mod.rs:
```diff
@@ -238,6 +239,30 @@ impl FileScanConfig {
 }
 
+/// The base configurations to provide when creating a physical plan for
+/// any given file format.
```
Done, thanks for reading in detail

tustvold commented on datafusion/core/src/physical_plan/file_format/mod.rs:
```diff
@@ -238,6 +239,30 @@ impl FileScanConfig {
+/// The base configurations to provide when creating a physical plan for
+/// writing to any given file format.
+#[derive(Debug, Clone)]
+pub struct FileSinkConfig {
```
:+1:

tustvold commented on datafusion/core/src/datasource/file_format/csv.rs:
```diff
@@ -324,6 +349,196 @@ fn build_schema_helper(names: Vec<String>, types: &[HashSet<DataType>]) -> Schem
+        for file_group in &self.config.file_groups {
+            let header = self.has_header
+                && (!matches!(&self.config.writer_mode, FileWriterMode::Append)
+                    || file_group.object_meta.size == 0);
```
For put and put multipart we should write the header regardless of any existing contents (as they will be overwritten)?

mustafasrepo replied:
Exactly, I have simplified the condition

tustvold reviewed (commented):
Had a quick review, I think this is probably fine, but type-erasing the writer mode seems a little peculiar to me.

This is because each of the methods has rather different characteristics, and imo warrants writing in a different manner.

## Put

The write is completely synchronous (it is writing to memory) and is then atomically flushed, with no need for abort behaviour or async write. All file formats can support this mode

## Put Multipart

The write is async with a final atomic close. Requires custom abort logic. All file formats can support this mode

## Append

Abort is fatal (not even entirely sure how to surface this), only supported by row oriented file formats. Even then requires custom handling for things like CSV headers, etc...

## Proposal

I guess my proposal would be to simply add a match block within `DataSink` for each of the various `FileWriterMode`. Over time I expect we will be able to extract common logic for each of the variants, e.g. a generic `Put` version using a `RecordBatchWriter`, etc... but I'm not sure that trying to unify all the writer modes is tractable and certainly at this stage where we only have one impl seems a touch premature perhaps

mustafasrepo commented on datafusion/core/src/physical_plan/file_format/csv.rs:
```diff
@@ -323,6 +328,78 @@ impl FileOpener for CsvOpener {
+    async fn create_writer(&self, file_meta: FileMeta) -> Result<Box<dyn FileWriterExt>> {
+        let object = &file_meta.object_meta;
+        match self.writer_mode {
+            // If the mode is append, call the store's append method and return a ready poll
+            // with the result wrapped in a custom error type if it fails
```
Fixed it

mustafasrepo commented on datafusion/core/src/datasource/file_format/mod.rs:
```diff
@@ -87,6 +98,277 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
+    }
+}
+
+/// A wrapper around an `AsyncWrite` type that provides append functionality.
```
You are right. These types were exactly same (we missed it and tried to replicate file writer mode enum). However, after suggestion of @tustvold I have changed the abort implementation of the append case.  Hence this is no longer the case. However, these type have common code, I will try to increase co reuse between these types so that the places they differ is more clear.

mustafasrepo commented:
> Had a quick review, I think this is probably fine, but type-erasing the writer mode seems a little peculiar to me.
> 
> This is because each of the methods has rather different characteristics, and imo warrants writing in a different manner.
> 
> ## Put
> The write is completely synchronous (it is writing to memory) and is then atomically flushed, with no need for abort behaviour or async write. All file formats can support this mode
> 
> ## Put Multipart
> The write is async with a final atomic close. Requires custom abort logic. All file formats can support this mode
> 
> ## Append
> Abort is fatal (not even entirely sure how to surface this), only supported by row oriented file formats. Even then requires custom handling for things like CSV headers, etc...
> 
> ## Proposal
> I guess my proposal would be to simply add a match block within `DataSink` for each of the various `FileWriterMode`. Over time I expect we will be able to extract common logic for each of the variants, e.g. a generic `Put` version using a `RecordBatchWriter`, etc... but I'm not sure that trying to unify all the writer modes is a good abstraction and certainly at this stage where we only have one impl seems a touch premature perhaps?

You proposal makes sense to me. I have removed `FileWriterFactory`. Now writer is created in the `CsvSink`. I in the future need for `trait` arises we can do so.

tustvold reviewed (approved):
Thank you, I would have gone further and removed FileWriterExt in favour of just a type-erased BatchSerializer, but we can always continue to iterate on this design

mustafasrepo commented:
> Thank you, I would have gone further and removed FileWriterExt in favour of just a type-erased BatchSerializer, but we can always continue to iterate on this design

I have removed `FileWriterExt`. We now use `AbortableWrite<W>`. This increases the code reuse. Thanks for the suggestion.

alamb commented on datafusion/core/src/datasource/file_format/mod.rs:
```diff
@@ -205,136 +205,74 @@ impl AsyncWrite for AsyncPutWriter {
-/// A wrapper around an `AsyncWrite` type that provides append functionality.
-pub struct AsyncAppend<W: AsyncWrite + Unpin + Send> {
+/// A wrapper struct with abort method and writer
+struct AbortableWrite<W: AsyncWrite + Unpin + Send> {
```
👍  I think this construction makes the intent of the code much clearer. Thank you

alamb commented:
I am going to merge this PR so we can continue work on main. I am really excited to see this progress. Thank you @mustafasrepo  @metesynnada @ozankabak  and @tustvold

alamb commented:
I filed https://github.com/apache/arrow-datafusion/issues/6569 to start tracking next steps for the streaming write API support

mustafasrepo commented on datafusion/core/src/physical_plan/file_format/mod.rs:
````diff
@@ -287,41 +311,53 @@ struct FileGroupsDisplay<'a>(&'a [Vec<PartitionedFile>]);
+///
+/// Prints in the format:
+/// ```text
+/// [file1, file2,...]
````
You are right, I think functionality should match document, and display `...` after a limit. 
I think during refactor, I have lost this functionality. But I am not sure. We can add a test for this use case, to not lose functionality in the future.

qrilka replied:
Will you have time to do this or I should try?

mustafasrepo replied:
I won't be able to look into this 2-3 days. In the mean time, if you want to do this, you can do so.

qrilka replied:
Thanks, I'll try to find a couple of hours over the weekend, will ping you if there will be something to share

qrilka replied:
PR created - https://github.com/apache/arrow-datafusion/pull/6637
